README.md
# bayesLopod
###### Bayesian Inference of Landscape Occupancy from Presence-Only Data

Natural history museums and herbaria collectively hold hundreds of millions of zoological, botanical, and paleontological specimens. These collections serve as the foundation for understanding the distribution of life on Earth and the basis for addressing loss of biodiversity, emerging diseases, and other pressing global problems as well as important question in ecology and evolution. One of the short comings of these kind of data is that the lack of evidence of the presence of a species in a certain region does not mean the species is truly absent there. Likewise, specimens are often misidentified, and therefore the report of a species in a locality is not always evidence that a viable population occurs there. The goal of this project is to develop a method which could be used to estimate the probability of presence of a species in sampling units of a study region based on certain sampling effort and presence detections pattern.

***

## Model

### `bayesLopod` as an occupancy model

[BRIEF EXPLANATION OF OCCUPANCY MODELS, VARIATION ACROSS SAMPLING EVENTS USING BENULLI PROCESSES]

`bayesLopod` assumes that on average, detectability is the same in a sampling unit, and therefore the probability of observing a certain sampling pattern (i.e. number of detection given a number of sampling events; `theta`), if the species occurs in a sampling unit, follows a binomial distribution:   

![Binomial Expanded](/gif/binomial_exp_eq.gif)
 in Stan, the log-probability of `y` detections, given `N` sampling events and, `p` detectability for each sampling unit, can be added to the target log-probability in the model block:

``` Stan
target  +=  binomial_lpmf(y | N, p);
```
#### Detectability

Because `bayesLopod` is designed to for sampling events that have not been systematically designed, detectability (`p`) has a broader interpretation than in traditional occupancy models.  In this case, `p` does not only account for the imperfect capacity of a researcher to detect a species he or she is looking for, but for other complexities of using non systematically collected presence only data. For example `p` would be decreased by reporting bias, in case that a researcher detects an individual of the species the species, but for whatever reason it is not registered in the database used (e.g. is a common species, other individuals have been detected previously, or it is not a species of interest for the collector) as well as the fact that a species might not occupy the entirety of a sampling unit (sampling units that include the range limits, would have smaller `p`).

Additionally, `bayesLopod` allow flexibility in what `N` represents, and it does not need to be a "sampling event", if this information is not available. `N`, for example, can be the number of detections of individuals in a "target group" (a group of species that a researcher is as likely to collect as the focal species): in this case, `p` represents the ratio of records of the focal species `y` and records of all species in the target group `N`.

#### (Site-level) False positives

Because presence-only data are often not collected by the researcher performing the `bayesLopod` model, there is limitations assessing the quality of every record. To allow flexibility in this regard, `bayesLopod` includes a rate of false negatives as a parameter `q`. This parameter can be estimated by the models or defined by the user (for example, if the user is confident there are no false positives, `q` can be set to 0).

It is worth noting that false positives are assessed at the "sampling-unit" level and not at the "record" level. This means that whether a detection is a false positive or not, depends only on the species occurrence in a sampling unit. Importantly, this will leave deem as true detections a set of false detections (those that were made in sampling units in which the species occurs) but will have no effect in the geographical range occupied.

Like probability of detection (`p`), the rate of false positives `q` has a broader interpretation than purely misidentified and mislabeled records. In `bayesLopod`, `q` would also include individuals that are indeed member of the focal species, but not in its main geographical range, for example vagrants, migrants, or invasive species. More research is needed to determine the way range dynamics interact with this parameter.

Finally, the probability of observing a certain sampling pattern (i.e. number of detection given a number of sampling events; `theta`), if the species **does not** occur in a sampling unit, follows a binomial distribution, with `q` as the "success" rate.  A key distinction with traditional occupancy models, is that to  account for false detections, they require a set of observations that are certain. However, `bayesLopod` estimates the rate of false detections based on two assumptions:

- The rate of false positives `q` is the same across all sampling units.
- The rate of true detections `p` is greater than that of  false positives `q`

The prior distribution of `qRate` (`q` / `p`) is:


 ![qRate Prior](/gif/qRate_prior.gif)

And is included in the model block of the Stan script as:
``` Stan
  target += normal_lpdf(qRate | 0, 0.05);
```

#### Probability of Occupancy (psi)

For each sampling unit, `bayesLopod` estimate a probability of occurrence (psi) maximizing the probability of a sampling pattern `thata` given a detection rate `p` and false positive rate `q`:

![Double Binomial](/gif/binomialpq_eq.gif)

Which is included in the `model` block of the Stan model.

``` Stan
target  += log_mix( psi_Sampled[cell],
                   binomial_lpmf(y[cell] | N[cell] , p),
                   binomial_lpmf(y[cell] | N[cell] , q)
                   );
```

Finally, the prior distribution of psi follows a Beta distribution with larger probabilities for values around 0 or 1.

![psi Prior](/gif/psi_prior.gif)
``` Stan
  target += binomial_lpdf(psy | 0.5, 0.5);
```
#### Variable detection rates across sampling units

The examples above include models in which detectability `p` is the same across all sampling units. However, `bayesLopod` allow for variation in detectability by assigning a `p` rate for each of the sampling units:

![Double Binomial pi](/gif/binomialpiq_eq.gif)

 To do so, it first determines a range (between `pmax` and `pmin`), and then determines a number between 0 and 1 for each of the sampling units (`praw`). If `praw` = 0, then the `p` value for that cell is `pmin`, and if `praw` = 1,  the `p` value for that cell is `pmax`. This is calculated in the `transformed parameters` block in the Stan scripts in which `p` varies:

``` Stan
  pRange = pmax - pmin;
  p = (p_raw * pRange) + pmin ;
```

Then, the 'model' block in these scripts is modified slightly to reflect the variable rate (note that `p[cell]` refers to a sampling unit value, whereas `q` is global)

``` Stan
  target  += log_mix( psi_Sampled[cell],
                      binomial_lpmf(y[cell] | N[cell] , p[cell]),
                      binomial_lpmf(y[cell] | N[cell] , q)
                      );
```

Lastly, `pRange` and `praw` have priors as follows:

- `pRange` around 0, making `pmin` and `pmax` close to each other.
- `praw` around 1, making most values of `p` close to `pmax`

| ![pRange Prior](/gif/pRange_prior.gif) | ![Praw Prior](/gif/praw_prior.gif) |
|---|---|


``` Stan
  target += normal_lpdf(pRange | 0, 0.1);
  target += normal_lpdf(p_raw  | 1, 0.25);
```

While `pmin` and `pmin` have broadly non informative priors:

| ![pRange Prior](/gif/pmin_prior.gif) | ![Praw Prior](/gif/pmax_prior.gif) |
|---|---|

``` Stan
  target += normal_lpdf(pmin | 0.5, 0.25);
  target += normal_lpdf(pmax | 0.5, 0.25);
```
#### CAR model

`bayesLopod` allows unsampled or poorly sampled sampling units to borrow power to predict occupancy based on the patterns of its neighboring units (i.e. a cell is more likely to be occupied if it is surrounded by cells in which the species is present). We do so by implementing a  conditional autoregressive (CAR) model, in which psi is spatially autocorrelated. We thank  @mbjoseph for making the code for this analyses publically available  [(see github repository for more details about this model in Stan and CAR in general)](https://github.com/mbjoseph/CARstan).

In addition to the `function` block defining `sparse_car_lpdf` the followin lines are added to the `model` block in the Stan files for models in which CAR is used. 

``` Stan
  target += sparse_car_lpdf(psi_i | tau, alpha, W_sparse, D_sparse, lambda, n, W_n);
  target += gamma_lpdf(tau | 2, 2);
  // prior for alpha does not add to target, thus it is an uniform distribution between 0 and 1
```

## Examples
### Case Study: _Andropogon gerardii_ (Shape File)

``` R
install.packages("bayesLopod", dependencies = T)

library(bayesLopod)
packageVersion("bayesLopod")
```
Make sure the version installed is > 1.0.1, the most recent version of bayesLopod can be installed from github usung the `devtools` package:
``` R
devtools::install_github( repo = "bayesLopod",
                          username = "camilosanin")

library(bayesLopod)
packageVersion("bayesLopod")
```

``` R
data("Andropogon_shape", package = "bayesLopod")
androSEff = spplot(Andropogon_shape, zcol = c("sampEffort"), main = "Sampling Effort")
androDetect = spplot(Andropogon_shape, zcol = c("detections"), main = "Detections")
plot(androSEff, split = c(1,1,2,1), more = T)
plot(androDetect,  split = c(2,1,2,1), more = F)
```

![Andropogon-input](/gif/Andropogon_input.gif)

``` R
ld_Shape = shapeLopodData( Shapefile = Andropogon_shape,
                           fieldN = "sampEffort",
                           fieldY = "detections",  
                           Adjacency = T,
                           keepFields = F)

```

``` R
mLopodShape = modelLopod( LopodData = ld_Shape,
                          varP = T,
                          q = NULL,
                          pmin = 0,
                          CAR = T,
                          nChains = 3,
                          warmup = 400,
                          sampling = 100,
                          nCores =3)
```
``` R
lopodTrace( mLopodShape, inc_warmup = T)
```
![Andropogon-trace](/gif/Andropogon_trace.gif)

``` R
lopodDens(mLopodShape, c("q", "pmin", "pmax"))
```
![Andropogon-dens](/gif/Andropogon_dens.gif)

``` R
AndroShape = mLopodShape@LopodData@geoDataObject

psiShape = lopodShape(mLopodShape, "psi_i", extrapolate = T,  quant = 0.05)
AndroShape@data[,"psi_05"] = psiShape@data[,"psi_i"]

psiShape = lopodShape(mLopodShape, "psi_i", extrapolate = T,  quant = 0.5)
AndroShape@data[,"psi_50"] = psiShape@data[,"psi_i"]

psiShape = lopodShape(mLopodShape, "psi_i", extrapolate = T,  quant = 0.95)
AndroShape@data[,"psi_95"] = psiShape@data[,"psi_i"]

spplot( AndroShape,
        zcol = c("psi_05", "psi_50", "psi_95"),
        names.attr = c("Psi (5% quantile)", "Psi (median)", "Psi (95% quantile)"), main = "Occupancy (Psi)")
        )
```
![Andropogon-psi-spplot](/gif/Andropogon_psi_spplot.gif)

### Case Study: Simulated Species (Coordinates and Raster)
``` R
data("simSpRecords", package = "bayesLopod")
data("simSpSamplingEffort", package = "bayesLopod")
```
``` R
simSpRasters = xyToRaster( xyRecords = simSpRecords,x
                           ySamplingEffort = simSpSamplingEffort,
                           basemap = NULL,
                           nrows = 50,
                           extentExpansion = 0)

spplot(simSpRasters, names.attr	=c("Species Detections","Sampling Effort"))
```
![XY-input](/gif/XY_Input.gif)
``` R
ld_Raster_adMatrix = rasterLopodData( rasterN = simSpRasters[["samplingEffort"]],
                                      rasterY = simSpRasters[["spDetections"]],
                                      Adjacency = T,
                                      extSample = 1.0,
                                      extDetection = 1.0 )

```
``` R
mLopodRaster = modelLopod( LopodData = ld_Raster_adMatrix,
                           varP = F,
                           q = NULL,
                           pmin = 0,
                           CAR = T,
                           nChains = 4,
                           warmup = 100,
                           sampling = 50,
                           nCores = 4 )
```

``` R
lopodTrace( mLopodRaster, inc_warmup = T)
```
![XY-trace](/gif/XYTrace.gif)
``` R
lopodDens(mLopodRaster, c("q", "p"))
```
![XY-dens](/gif/XYDens.gif)
```R
ppRaster = lopodRaster(mLopodRaster, param = "pp", extrapolate = F, metric = "mean")
psiRaster = lopodRaster(mLopodRaster, param = "psi_i", extrapolate = T, metric = "mean")
```

![Probability of presence](/gif/prPres_eq.gif)
``` R
spplot( raster::stack(psiRaster,ppRaster),  
        names.attr = c("Occupancy (Psi)", "Probability of Presence"))
```
![XY-psi-spplot](/gif/XY_raster.gif)


NEWS.md
bayesLopod 2.0.0

Rewritten using JAGS!

bayesLopod 1.0.2

More detailed documentation, particularly in modeLopod.
rhat can be plotted in shape and rasters

bayesLopod 1.0.1

Correct spelling of Psi across all package scripts.
Correction in priors for stanfiles with variable detection probabilities (pi).


R/modelLopod.R
#' Run a  Bayes inference of Landscape Occupancy from Presence-Only Data on Stan
#' @useDynLib bayesLopod, .registration = TRUE
#' @import Rcpp
#' @import rstan
#' @import rstantools
#' @param LopodData \linkS4class{LopodData} with the data to be used in the Model
#' @param varP Boolean. If TRUE, detectability will vary across cells. If FALSE a global value for detectability will be estimated.
#' @param CAR Boolean. If TRUE, (and if a adjacency matrix is included in the LopodData object) a conditional auto-regression analysis will be performed for occupancy across units.
#' @param pmin Number between 0 and 1. Minimum value for detectability in a unit in which the species occurs.
#' @param q Number between 0 and 1 or NULL. Rate of false detections. If NULL the values will be estimated by the model.
#' @param nChains Number of Markov chains used by the Stan model.
#' @param warmup Number of iterations for each chain to be discarded as warm-up.
#' @param sampling Number of iterations for each chain to be sampled (after warm-up).
#' @param nCores Number of cores to use when executing the chains in parallel.
#' @export
#' @return A \linkS4class{LopodModel} object.
#' @details
#' These parameters are estimated for the entire sampling area and affect all sampling units equally. These parameters can be examined using the \code{\link{lopodDens}}, \code{\link{lopodSummary}} and \code{\link{lopodTrace}} functions in bayesLopod.
#' \itemize{
#'   \item "psi": Average occupancy across the study area.
#'   \item "p": (if varP = FALSE) Probability of detection for all sampling units.
#'   \item "pmax": (if varP = TRUE) Maximum value of p.
#'   \item "pmin": (if varP = TRUE) Minimum value of p.
#'   \item "alpha": (if CAR = TRUE) alpha parameter of the conditional autoregressive model.
#'   \item "tau": (if CAR = TRUE) tau parameter of the conditional autoregressive model.
#'   \item "chi_sq"*: (EXPERIMENTAL) chi-square statistic for model.
#'   \item "lLh"*: (EXPERIMENTAL) total log-likelihood for the pattern observed across all sampling units, given the parameters estimated in the model.
#'   \item "AIC"*: (EXPERIMENTAL) Akaike information criterion value for the model.
#' }
#'
#' \strong{Sampled sampling-units}:
#' These parameters are estimated for each sampling unit that has at least one sampling event. These parameters can be retrieved to a Shapefile or Raster using the \code{\link{lopodRaster}} or \code{\link{lopodShape}}functions.
#' \itemize{
#'   \item "psi_Sampled": Probability of occurrence in each sampling unit.
#'   \item "pCorr": (if varP = TRUE) probability of detection in each sampling unit that is occupied. This value is corrected to include unoccupied cells, in which case p = 0.
#'   \item "sim_y": Simulated number of detections based on the number of sampling events and the model.
#'   \item "sim_true_y": Simulated number of true detections based on the number of sampling events and the model.
#'   \item "sim_false_y": Simulated number of false detections based on the number of sampling events and the model.
#'   \item "pp": Probability of presence given a pattern of sampling effort/detections and the model.
#'   \item "cellpres_i": Simulated presence/absence for each iteration based on model.
#'   \item "expRec"*: (EXPERIMENTAL) Expected number of detections.
#'   \item "lLh_cell"*: (EXPERIMENTAL) Likelihood of the pattern in a cell given the model.
#' }
#'
#' \strong{All sampling-units}:
#' These parameters are estimated for all sampling units only if a conditional autoregressive model  is implemented (CAR = TRUE). These parameters can be retrieved to a Shapefile or Raster using the \code{\link{lopodRaster}} or \code{\link{lopodShape}} functions.
#' \itemize{
#'   \item "psi_i":  Probability of occurrence in each sampling units. This is inferred for unsampled united using the CAR model. For sampled units, "psi_i" equals "psi_Sampled"
#' }
#' *EXPERIMENTAL VALUES SHOULD BE TAKEN CAREFULLY AND NOT USED NAIVELY TO DETERMINE WHICH MODELS ARE BETTER THAN OTHERS
#' @examples
#' data("simSpRecords", package = "bayesLopod")
#' data("simSpSamplingEffort", package = "bayesLopod")
#' simSpRasters = xyToRaster(xyRecords = simSpRecords,xySamplingEffort = simSpSamplingEffort,
#' basemap = NULL, nrows = 10, extentExpansion = 0)
#' ld_Raster = rasterLopodData(rasterN = simSpRasters[["samplingEffort"]],
#' rasterY = simSpRasters[["spDetections"]], Adjacency = FALSE )
#' mLopodRaster = modelLopod(LopodData = ld_Raster, varP = TRUE, q = NULL,
#' pmin = 0, CAR = FALSE, nChains = 1,warmup = 10,sampling = 10,nCores = 1)
#'
#' data("Andropogon_shape", package = "bayesLopod")
#' ld_Shape = shapeLopodData(Shapefile = Andropogon_shape, fieldN = "sampEffort",
#' fieldY = "detections",  Adjacency = FALSE, keepFields = FALSE)
#' mLopodShape = modelLopod(LopodData = ld_Shape, varP = FALSE, q = NULL,
#' pmin = 0, CAR = FALSE, nChains = 1,warmup = 5,sampling = 5,nCores = 1)
#'
#' \dontrun{
#' data("simSpRecords", package = "bayesLopod")
#' data("simSpSamplingEffort", package = "bayesLopod")
#' simSpRasters = xyToRaster(xyRecords = simSpRecords,xySamplingEffort = simSpSamplingEffort,
#' basemap = NULL, nrows = 50, extentExpansion = 0)
#' ld_Raster_adMatrix = rasterLopodData(rasterN = simSpRasters[["samplingEffort"]],
#' rasterY = simSpRasters[["spDetections"]], Adjacency = TRUE )
#' mLopodRaster = modelLopod(LopodData = ld_Raster_adMatrix, varP = TRUE, q = NULL,
#' pmin = 0.1, CAR = FALSE, nChains = 4,warmup = 500,sampling = 100,nCores = 4)
#'
#' data("Andropogon_shape", package = "bayesLopod")
#' ld_Shape = shapeLopodData(Shapefile = Andropogon_shape, fieldN = "sampEffort",
#' fieldY = "detections",  Adjacency = TRUE, keepFields = FALSE)
#' mLopodShape = modelLopod(LopodData = ld_Shape, varP = TRUE, q = NULL,
#' pmin = 0, CAR = TRUE, nChains = 4,warmup = 500,sampling = 100,nCores =4)
#' }


modelLopod = function(LopodData, varP = F, q =  NULL, pmin = 0, CAR = F, nChains = 4, warmup = 2000, sampling = 1000, nCores=4){

  stanFilesDir = "./exec/"

#Check values anda data types

if (class(LopodData) != "LopodData") stop("Data should be entered in a LopodData Object")

if (is.null(q)==F){
  if (q<0 | q>1) stop("q must be NULL or between 0 and 1")
}

if (pmin<0 | pmin>1) stop("pmin must be between 0 and 1")

#Get N and y from different formats

if(LopodData@geoType == "Raster"){

  N = LopodData@geoDataObject[["samplingEffort"]][LopodData@geoInfo$sampledId$cellRaster]
  y = LopodData@geoDataObject[["Detections"]][LopodData@geoInfo$sampledId$cellRaster]

}
if(LopodData@geoType == "Shapefile"){

  N = LopodData@geoDataObject$sampEffort[LopodData@geoInfo$sampledId$featureShape]
  y = LopodData@geoDataObject$detections[LopodData@geoInfo$sampledId$featureShape]

}

  if(LopodData@geoType != "Shapefile"&LopodData@geoType != "Raster"){
    stop("Only Raster or Shapefiles formats are currently supported")}

  modelInfo = list(varP = varP, q =  q, CAR = CAR)

  #### Stan Files and Stan Data

  if (CAR == F) {

    if (varP == F){

      if (is.null(q)==T){


        if(LopodData@geoType == "Raster"){
          stanData = list( nSampledCells = length(LopodData@geoInfo$sampledId$cellRaster),
                           sampledId = LopodData@geoInfo$sampledId$cellStan,
                           N = N,
                           y = y,
                           minP = pmin
          )

        }
        if(LopodData@geoType == "Shapefile"){
          stanData = list( nSampledCells = length(LopodData@geoInfo$sampledId$featureShape),
                           sampledId = LopodData@geoInfo$sampledId$cellStan,
                           N = N,
                           y = y,
                           minP = pmin
          )

        }


        message("Global p and q estimated. Psi for each sampling unit.")
        StanModel = stanmodels$psiipq


      }

      if (is.null(q)==F) {

        if (pmin < q) pmin = q

        if(LopodData@geoType == "Raster"){

          stanData = list( nSampledCells = length(LopodData@geoInfo$sampledId$cellRaster),
                           sampledId = LopodData@geoInfo$sampledId$cellStan,
                           N = N,
                           y = y,
                           minP = pmin,
                           q = q

          )

        }
        if(LopodData@geoType == "Shapefile"){

          stanData = list( nSampledCells = length(LopodData@geoInfo$sampledId$featureShape),
                           sampledId = LopodData@geoInfo$sampledId$cellStan,
                           N = N,
                           y = y,
                           minP = pmin,
                           q = q

          )

        }


        message("Global p estimated assuming p is larger than the given q (which can be 0, in which case there are no false detections). Psi estimated for each sampling unit.")
        StanModel = stanmodels$psiip

      }
    }

    if (varP == T){

      if (is.null(q)==T){

        if(LopodData@geoType == "Raster"){

          stanData = list( nSampledCells = length(LopodData@geoInfo$sampledId$cellRaster),
                           sampledId = LopodData@geoInfo$sampledId$cellStan,
                           N = N,
                           y = y,
                           minP = pmin
          )

        }
        if(LopodData@geoType == "Shapefile"){

          stanData = list( nSampledCells = length(LopodData@geoInfo$sampledId$featureShape),
                           sampledId = LopodData@geoInfo$sampledId$cellStan,
                           N = N,
                           y = y,
                           minP = pmin
          )

        }


        message("Global q estimated. Psi and P for each sampling unit")

          StanModel = stanmodels$psiipiq


      }

      if (is.null(q)==F) {

        if (pmin < q) pmin = q

        if(LopodData@geoType == "Raster"){
          stanData = list( nSampledCells = length(LopodData@geoInfo$sampledId$cellRaster),
                           sampledId = LopodData@geoInfo$sampledId$cellStan,
                           N = N,
                           y = y,
                           minP = pmin,
                           q = q
          )



        }
        if(LopodData@geoType == "Shapefile"){

          stanData = list( nSampledCells = length(LopodData@geoInfo$sampledId$featureShape),
                           sampledId = LopodData@geoInfo$sampledId$cellStan,
                           N = N,
                           y = y,
                           minP = pmin,
                           q = q
          )



        }

        message("Psi and p estimated for each sampling unit assuming p is larger than the given q (which can be 0, in which case there are no false detections).")
        StanModel = stanmodels$psiipi


      }
    }
  }

  if (CAR == T) {
    if(is.null(LopodData@geoInfo$W_sparse)==T){
      stop("Cannot perform CAR analysis without adjacency matrix")
    } else {

      if (varP == F){

        if (is.null(q)==T){

           if(LopodData@geoType == "Raster"){
             stanData = list( nSampledCells = length(LopodData@geoInfo$sampledId$cellRaster),
                              sampledId = LopodData@geoInfo$sampledId$cellStan,
                              nNotSampled = length(LopodData@geoInfo$notSampledId$cellStan),
                              notSampledId = LopodData@geoInfo$notSampledId$cellStan,
                              n = length(LopodData@geoInfo$sampledId$cellRaster)+length(LopodData@geoInfo$notSampledId$cellStan),
                              W_n = dim(LopodData@geoInfo$W_sparse)[1],
                              W_sparse = LopodData@geoInfo$W_sparse,
                              D_sparse = LopodData@geoInfo$D_sparse,
                              lambda = LopodData@geoInfo$lambda_sparse,
                              N = N,
                              y = y,
                              minP = pmin
             )

          }
          if(LopodData@geoType == "Shapefile"){

            stanData = list( nSampledCells = length(LopodData@geoInfo$sampledId$featureShape),
                             sampledId = LopodData@geoInfo$sampledId$cellStan,
                             nNotSampled = length(LopodData@geoInfo$notSampledId$cellStan),
                             notSampledId = LopodData@geoInfo$notSampledId$cellStan,
                             n = length(LopodData@geoInfo$sampledId$featureShape)+length(LopodData@geoInfo$notSampledId$cellStan),
                             W_n = dim(LopodData@geoInfo$W_sparse)[1],
                             W_sparse = LopodData@geoInfo$W_sparse,
                             D_sparse = LopodData@geoInfo$D_sparse,
                             lambda = LopodData@geoInfo$lambda_sparse,
                             N = N,
                             y = y,
                             minP = pmin
            )

          }


          message("Global p and q estimated. Psi for each sampling unit. Psi is spatially auto-correlated.")
          StanModel = stanmodels$psiipq_CAR

        }

        if (is.null(q)==F) {

          if (pmin < q) pmin = q

          if(LopodData@geoType == "Raster"){

            stanData = list( nSampledCells = length(LopodData@geoInfo$sampledId$cellRaster),
                             sampledId = LopodData@geoInfo$sampledId$cellStan,
                             nNotSampled = length(LopodData@geoInfo$notSampledId$cellStan),
                             notSampledId = LopodData@geoInfo$notSampledId$cellStan,
                             n = length(LopodData@geoInfo$sampledId$cellRaster)+length(LopodData@geoInfo$notSampledId$cellStan),
                             W_n = dim(LopodData@geoInfo$W_sparse)[1],
                             W_sparse = LopodData@geoInfo$W_sparse,
                             D_sparse = LopodData@geoInfo$D_sparse,
                             lambda = LopodData@geoInfo$lambda_sparse,
                             N = N,
                             y = y,
                             minP = pmin,
                             q = q
            )


          }
          if(LopodData@geoType == "Shapefile"){

            stanData = list( nSampledCells = length(LopodData@geoInfo$sampledId$featureShape),
                             sampledId = LopodData@geoInfo$sampledId$cellStan,
                             nNotSampled = length(LopodData@geoInfo$notSampledId$cellStan),
                             notSampledId = LopodData@geoInfo$notSampledId$cellStan,
                             n = length(LopodData@geoInfo$sampledId$featureShape)+length(LopodData@geoInfo$notSampledId$cellStan),
                             W_n = dim(LopodData@geoInfo$W_sparse)[1],
                             W_sparse = LopodData@geoInfo$W_sparse,
                             D_sparse = LopodData@geoInfo$D_sparse,
                             lambda = LopodData@geoInfo$lambda_sparse,
                             N = N,
                             y = y,
                             minP = pmin,
                             q = q
            )


          }



          message("Global p estimated assuming p is larger than the given q (which can be 0, in which case there are no false detections). Psi estimated for each sampling unit. Psi is spatially auto-correlated.")
          StanModel = stanmodels$psiip_CAR

        }
      }

      if (varP == T){

        if (is.null(q)==T){

          if(LopodData@geoType == "Raster"){

            stanData = list( nSampledCells = length(LopodData@geoInfo$sampledId$cellRaster),
                             sampledId = LopodData@geoInfo$sampledId$cellStan,
                             nNotSampled = length(LopodData@geoInfo$notSampledId$cellStan),
                             notSampledId = LopodData@geoInfo$notSampledId$cellStan,
                             n = length(LopodData@geoInfo$sampledId$cellRaster)+length(LopodData@geoInfo$notSampledId$cellStan),
                             W_n = dim(LopodData@geoInfo$W_sparse)[1],
                             W_sparse = LopodData@geoInfo$W_sparse,
                             D_sparse = LopodData@geoInfo$D_sparse,
                             lambda = LopodData@geoInfo$lambda_sparse,
                             N = N,
                             y = y,
                             minP = pmin
            )


          }

          if(LopodData@geoType == "Shapefile"){

            stanData = list( nSampledCells = length(LopodData@geoInfo$sampledId$featureShape),
                             sampledId = LopodData@geoInfo$sampledId$cellStan,
                             nNotSampled = length(LopodData@geoInfo$notSampledId$cellStan),
                             notSampledId = LopodData@geoInfo$notSampledId$cellStan,
                             n = length(LopodData@geoInfo$sampledId$featureShape)+length(LopodData@geoInfo$notSampledId$cellStan),
                             W_n = dim(LopodData@geoInfo$W_sparse)[1],
                             W_sparse = LopodData@geoInfo$W_sparse,
                             D_sparse = LopodData@geoInfo$D_sparse,
                             lambda = LopodData@geoInfo$lambda_sparse,
                             N = N,
                             y = y,
                             minP = pmin
            )


          }


          message("Global q estimated. Psi and P for each sampling unit. Psi is spatially auto-correlated.")
          StanModel = stanmodels$psiipiq_CAR

        }

        if (is.null(q)==F) {

          if (pmin < q) pmin = q

          if(LopodData@geoType == "Raster"){

            stanData = list( nSampledCells = length(LopodData@geoInfo$sampledId$cellRaster),
                             sampledId = LopodData@geoInfo$sampledId$cellStan,
                             nNotSampled = length(LopodData@geoInfo$notSampledId$cellStan),
                             notSampledId = LopodData@geoInfo$notSampledId$cellStan,
                             n = length(LopodData@geoInfo$sampledId$cellRaster)+length(LopodData@geoInfo$notSampledId$cellStan),
                             W_n = dim(LopodData@geoInfo$W_sparse)[1],
                             W_sparse = LopodData@geoInfo$W_sparse,
                             D_sparse = LopodData@geoInfo$D_sparse,
                             lambda = LopodData@geoInfo$lambda_sparse,
                             N = N,
                             y = y,
                             minP = pmin,
                             q = q
            )


          }
          if(LopodData@geoType == "Shapefile"){

            stanData = list( nSampledCells = length(LopodData@geoInfo$sampledId$featureShape),
                             sampledId = LopodData@geoInfo$sampledId$cellStan,
                             nNotSampled = length(LopodData@geoInfo$notSampledId$cellStan),
                             notSampledId = LopodData@geoInfo$notSampledId$cellStan,
                             n = length(LopodData@geoInfo$sampledId$featureShape)+length(LopodData@geoInfo$notSampledId$cellStan),
                             W_n = dim(LopodData@geoInfo$W_sparse)[1],
                             W_sparse = LopodData@geoInfo$W_sparse,
                             D_sparse = LopodData@geoInfo$D_sparse,
                             lambda = LopodData@geoInfo$lambda_sparse,
                             N = N,
                             y = y,
                             minP = pmin,
                             q = q
            )


          }

          message("Psi and p estimated for each sampling unit assuming p is larger than the given q (which can be 0, in which case there are no false detections). Psi is spatially auto-correlated.")
          StanModel = stanmodels$psiipi_CAR

        }
      }
    }
  }


 modelArgs = list(object = StanModel,
           data = stanData,              # named list of data
           chains = nChains,                   # number of Markov chains
           warmup = warmup,               # number of warmup iterations per chain
           iter = sampling+warmup,                 # total number of iterations per chain
           cores = nCores,                    # number of cores
           refresh = 2,                 # show progress every 'refresh' iterations
           control = list(adapt_delta  = 0.8)
  )

 StanFittetModel = do.call("sampling", args = modelArgs, envir = .GlobalEnv)

  return(LopodModel(LopodData = LopodData, StanFit = StanFittetModel, modelInfo = modelInfo))
}
